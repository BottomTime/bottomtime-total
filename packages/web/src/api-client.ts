import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
