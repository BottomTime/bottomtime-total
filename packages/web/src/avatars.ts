import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
