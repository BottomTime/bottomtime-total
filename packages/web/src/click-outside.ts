import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
