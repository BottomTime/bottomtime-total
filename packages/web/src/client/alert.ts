import { AlertDTO, CreateOrUpdateAlertParamsSchema } from '@bottomtime/api';

import { AxiosInstance } from 'axios';

export class Alert {
  constructor(
    private readonly client: AxiosInstance,
    private readonly data: AlertDTO,
  ) {}

  get id(): string {
    return this.data.id;
  }

  get icon(): string {
    return this.data.icon;
  }
  set icon(value: string) {
    this.data.icon = value;
  }

  get title(): string {
    return this.data.title;
  }
  set title(value: string) {
    this.data.title = value;
  }

  get message(): string {
    return this.data.message;
  }
  set message(value: string) {
    this.data.message = value;
  }

  get active(): Date {
    return this.data.active;
  }
  set active(value: Date) {
    this.data.active = value;
  }

  get expires(): Date | undefined {
    return this.data.expires;
  }
  set expires(value: Date | undefined) {
    this.data.expires = value;
  }

  toJSON(): AlertDTO {
    return { ...this.data };
  }

  async save(): Promise<void> {
    await this.client.put(
      `/api/alerts/${this.data.id}`,
      CreateOrUpdateAlertParamsSchema.parse(this.data),
    );
  }

  async delete(): Promise<void> {
    await this.client.delete(`/api/alerts/${this.data.id}`);
  }
}
