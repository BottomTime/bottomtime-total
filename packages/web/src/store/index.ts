export * from './alerts.store';
export * from './current-user.store';
export * from './operators.store';
export * from './dive-sites.store';
export * from './error.store';
export * from './friends.store';
export * from './log-entries.store';
export * from './profiles.store';
export * from './tanks.store';
export * from './toasts.store';
