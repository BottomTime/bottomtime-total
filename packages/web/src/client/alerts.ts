import {
  AlertSchema,
  CreateOrUpdateAlertParamsDTO,
  ListAlertsParamsDTO,
  ListAlertsResponseSchema,
} from '@bottomtime/api';

import { AxiosInstance } from 'axios';

import { Alert } from './alert';

export class AlertsApiClient {
  constructor(private readonly apiClient: AxiosInstance) {}

  async listAlerts(options?: ListAlertsParamsDTO): Promise<{
    alerts: Alert[];
    totalCount: number;
  }> {
    const { data } = await this.apiClient.get('/api/alerts', {
      params: options,
    });

    const parsed = ListAlertsResponseSchema.parse(data);
    return {
      alerts: parsed.alerts.map((alert) => new Alert(this.apiClient, alert)),
      totalCount: parsed.totalCount,
    };
  }

  async getAlert(alertId: string): Promise<Alert> {
    const { data } = await this.apiClient.get(`/api/alerts/${alertId}`);
    return new Alert(this.apiClient, AlertSchema.parse(data));
  }

  async createAlert(options: CreateOrUpdateAlertParamsDTO): Promise<Alert> {
    const { data } = await this.apiClient.post('/api/alerts', options);
    return new Alert(this.apiClient, AlertSchema.parse(data));
  }

  wrapDTO(data: unknown): Alert {
    return new Alert(this.apiClient, AlertSchema.parse(data));
  }
}
