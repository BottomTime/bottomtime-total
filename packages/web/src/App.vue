<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
