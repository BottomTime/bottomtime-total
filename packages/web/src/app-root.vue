<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
