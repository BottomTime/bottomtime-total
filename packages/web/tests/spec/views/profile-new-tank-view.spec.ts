import { ApiClient, TankDTO, TankMaterial } from '@bottomtime/api';
import { Tank } from '@bottomtime/api/src/client/tank';

import {
  ComponentMountingOptions,
  flushPromises,
  mount,
} from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';
import { Router } from 'vue-router';

import { ApiClientKey } from '../../../src/api-client';
import { LocationKey, MockLocation } from '../../../src/location';
import { useCurrentUser, useToasts } from '../../../src/store';
import ProfileNewTankView from '../../../src/views/profile-new-tank-view.vue';
import { createAxiosError } from '../../fixtures/create-axios-error';
import { createRouter } from '../../fixtures/create-router';
import {
  AdminUser,
  BasicUser,
  UserWithFullProfile,
} from '../../fixtures/users';

const DefaultUrl = `/profile/${BasicUser.username}/tanks/new`;

const TestData: TankDTO = {
  id: '3a3d7f74-f94f-4787-8f65-2636aee3ae7a',
  name: 'Freshest of All Tanks',
  isSystem: false,
  material: TankMaterial.Aluminum,
  volume: 11.6,
  workingPressure: 260,
};

describe('New Profile Tank view', () => {
  let client: ApiClient;
  let router: Router;

  let location: MockLocation;
  let pinia: Pinia;
  let currentUser: ReturnType<typeof useCurrentUser>;
  let toasts: ReturnType<typeof useToasts>;
  let opts: ComponentMountingOptions<typeof ProfileNewTankView>;

  beforeAll(() => {
    client = new ApiClient();
    router = createRouter([
      {
        path: '/profile/:username/tanks/new',
        component: ProfileNewTankView,
      },
    ]);
  });

  beforeEach(async () => {
    location = new MockLocation();
    location.assign(DefaultUrl);

    pinia = createPinia();
    currentUser = useCurrentUser(pinia);
    toasts = useToasts(pinia);
    currentUser.user = AdminUser;

    await router.push(DefaultUrl);

    opts = {
      global: {
        plugins: [router, pinia],
        provide: {
          [ApiClientKey as symbol]: client,
          [LocationKey as symbol]: location,
        },
      },
    };
  });

  it('will show the login form if the user is not logged in', async () => {
    currentUser.user = null;
    const wrapper = mount(ProfileNewTankView, opts);
    expect(wrapper.find('[data-testid="login-form"]').isVisible()).toBe(true);
    expect(wrapper.find('#name').exists()).toBe(false);
  });

  it("will show a forbidden message if the current user is not authorized to edit the target user's profile", async () => {
    currentUser.user = UserWithFullProfile;
    const wrapper = mount(ProfileNewTankView, opts);
    expect(wrapper.find('[data-testid="forbidden-message"]').isVisible()).toBe(
      true,
    );
    expect(wrapper.find('#name').exists()).toBe(false);
  });

  it('will allow a user to create a tank', async () => {
    const spy = jest
      .spyOn(client.tanks, 'createTank')
      .mockResolvedValue(new Tank(client.axios, TestData));

    const wrapper = mount(ProfileNewTankView, opts);
    await wrapper.get('#name').setValue(TestData.name);
    await wrapper.get('#material-al').setValue(true);
    await wrapper.get('#volume').setValue(TestData.volume);
    await wrapper.get('#pressure').setValue(TestData.workingPressure);
    await wrapper.get('#save-tank').trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledWith(
      {
        ...TestData,
        id: '',
      },
      BasicUser.username,
    );
    expect(location.pathname).toBe(
      `/profile/${BasicUser.username}/tanks/${TestData.id}`,
    );
  });

  it('will allow an admin to create a tank for another user', async () => {
    currentUser.user = AdminUser;
    const spy = jest
      .spyOn(client.tanks, 'createTank')
      .mockResolvedValue(new Tank(client.axios, TestData));

    const wrapper = mount(ProfileNewTankView, opts);
    await wrapper.get('#name').setValue(TestData.name);
    await wrapper.get('#material-al').setValue(true);
    await wrapper.get('#volume').setValue(TestData.volume);
    await wrapper.get('#pressure').setValue(TestData.workingPressure);
    await wrapper.get('#save-tank').trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledWith(
      {
        ...TestData,
        id: '',
      },
      BasicUser.username,
    );
    expect(location.pathname).toBe(
      `/profile/${BasicUser.username}/tanks/${TestData.id}`,
    );
  });

  it('will not attempt to create a tank if validation fails', async () => {
    const spy = jest.spyOn(client.tanks, 'createTank');
    const wrapper = mount(ProfileNewTankView, opts);
    await wrapper.get('#volume').setValue(0);
    await wrapper.get('#save-tank').trigger('click');
    await flushPromises();

    expect(
      wrapper.find('[data-testid="edit-tank-form-errors"]').text(),
    ).toMatchSnapshot();
    expect(spy).not.toHaveBeenCalled();
  });

  it('will render a toast message if the request fails because the user is at their tank limit', async () => {
    jest
      .spyOn(client.tanks, 'createTank')
      .mockRejectedValue(createAxiosError(405));
    const wrapper = mount(ProfileNewTankView, opts);

    await wrapper.get('#name').setValue(TestData.name);
    await wrapper.get('#material-al').setValue(true);
    await wrapper.get('#volume').setValue(TestData.volume);
    await wrapper.get('#pressure').setValue(TestData.workingPressure);
    await wrapper.get('#save-tank').trigger('click');
    await flushPromises();

    expect(toasts.toasts).toHaveLength(1);
    expect(toasts.toasts[0].id).toBe('tank-limit-reached');
    expect(toasts.toasts[0].message).toMatchSnapshot();
  });
});
