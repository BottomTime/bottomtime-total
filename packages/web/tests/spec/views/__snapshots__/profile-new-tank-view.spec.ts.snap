// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Profile Tank view will not attempt to create a tank if validation fails 1`] = `"There are errors in the form above. Please fix them and then try again. Name of tank profile is requiredVolume must be between 1 and 30 litresMax pressure is required"`;

exports[`New Profile Tank view will render a toast message if the request fails because the user is at their tank limit 1`] = `"Uh oh! It seems you have reached your limit for custom tank profiles. (Each user is limited to 10 custom tank profiles.) You can either edit an existing profile, or delete one and then create a new profile."`;
